### ISSUE TITLE
"Found error at index"

### ISSUE THREAD

**Issue by rivetti22** [wzygzlm/abmof_libcaer#1]
Hi there,

When I run the algorithm via "sudo ./abmof_libcaer", my terminal shows a lot of messages saying "Found error at index: ...".

![Screenshot from 2021-04-16 15-56-59](https://user-images.githubusercontent.com/38293419/115043429-6a5e1600-9ecc-11eb-8eaf-6165d92b1612.png)

I'm not sure of which event file is being read, but from my understanding, it is reading the file named "jAER-events.txt". Can you please confirm?
Other question. For the application to read my .txt files, do I need to pass any arguments when I run the application? Or do I need to change the source code?

Thank you in advance :smiley:

**Comment by wzygzlm** [comment]
Hi Alessio,

It has 5 parameters and all these parameters have default values.

![image](https://user-images.githubusercontent.com/8357338/115047613-f83ffe00-9ed8-11eb-885d-8381aa9d83d5.png)

The default value for the string name is box_trans-events.txt. Have you also downloaded this file?

**Comment by rivetti22** [comment]
Hi Min,

Okay, I didn't notice that it had these 5 parameters. 
Yes, I downloaded that file. Is it normal that it outputs those index errors?

Thank you for your quick response.

**Comment by wzygzlm** [comment]
Hi Alessio,

![image](https://user-images.githubusercontent.com/8357338/115052149-e44acb00-9edd-11eb-81e4-8d765ff763cf.png)

This is the error reported by result checking.  It means that the output OF result different from the Ground Truth.
But this might be caused by a new format for storing GT OF but I didn't update the code.

In a word, I think the algorithm should be fine, but the interpretation process of the GT data might have some bugs. So if you just want to try the C++ version of ABMOF, you don't need to care about it too much and you can comment on these lines.

**Comment by rivetti22** [comment]
Alright, since I'm not using GT I will comment those lines.

Thank you for your help 👍

**Comment by rivetti22** [comment]
Min, I have one more question.

The code seems to run indefinitely. I left it running for half an hour and it didn't stop. I used the "box_trans-events.txt" file.
I noticed that it started reading the file from the start, after it reached the end of the file. I believe the image below exposes exactly what I'm trying to explain.

![Screenshot from 2021-04-16 18-32-36](https://user-images.githubusercontent.com/38293419/115062360-2c6bec80-9ee2-11eb-8acd-979196d28b1f.png)

(My goal is to obtain a file like "box_trans-WithOFResult.txt", as obtained with the PatchMatchFlow in jAER.)
Once again, thank you for your help! :+1:

**Comment by wzygzlm** [comment]
Sorry. The purpose of this program is used to estimate OF result and the result is then sent to PC via TCP/UDP for display. Because it is supposed to run on an embedded device such as Raspberry pi. It doesn't save result data to a file. If you want to obtain a file as you mentioned in jAER, you need to modify the code. 

But my question is if you want to obtain that file, why not run in jAER directly?

**Comment by rivetti22** [comment]
Oh I see. That's why it defines a socket type. I should have noticed that.

Nevertheless, I didn't use jAER because I was looking for an implementation I could use with runtime performance. In jAER I need to manually start logging the OF values and I don't want that. But if I could extract the values automatically in jAER, that would be awesome. I just don't know how to do it, since I don't have experience working with Java coding/environments.

**Comment by rivetti22** [comment]
Hi Min,

Can you please tell me if it is possible to extract optical flow information in jAER (using PatchMatchFlow), without running the jAER in the JDK? 
In other words, can I run PatchMatchFlow in a standalone manner, to use in a real time application?

Thanks in advance.

**Comment by wzygzlm** [comment]
Hi Alessio,

That seems not possible. Are you gonna extract optical flow in real time? Like process optical flow for events directly from an event camera.

**Comment by rivetti22** [comment]
Hi Min,

Yes, that is my objective. Right now, I'm trying to edit your C++ ABMOF code to make it write the values to a file. 
Moreover, since I'm working with a CeleX event camera, I might explore the Optical Flow capabilities packed in its API.

**Comment by wzygzlm** [comment]
The code could process optical flow in an embedded device. It can accept event files but also a live event stream as input. But it is for DVS/DAVIS so the low-level API might differ from Celex camera.
 
![image](https://user-images.githubusercontent.com/8357338/115539775-ab27a780-a29d-11eb-9ffb-7c1533e1cd0a.png)
As shown in this figure, parseEventsSW is the ABMOF algorithm. The input of this function is the event frame and the output is a formatted event frame (including optical flow result). So you could save this output data into a file directly.

**Comment by rivetti22** [comment]
Yes, I believe that it would be different for CeleX. 
Nevertheless, I had already identified that "parseEventsSW" would return the data I wish to store into a file, as you said. I already extracted the data from the "eventSliceSW" variable. I only had issues obtaining the timestamp, but I was able to extract it from the "data" variable that the function receives as argument. 
Now I just need to store the variables in a file.

Thank you for all your help, I really appreciate it :smiley:
